cpp/src/arrow/util/bitmap_ops.cc

```diff
@@ -249,59 +249,92 @@ void AlignedBitmapOp(const uint8_t* left, int64_t left_offset, const uint8_t* ri
   left += left_offset / 8;
   right += right_offset / 8;
   out += out_offset / 8;
+  uint64_t outPopCount = 0;
   for (int64_t i = 0; i < nbytes; ++i) {
     out[i] = op(left[i], right[i]);
+    if (ComputeNewValidityCount) {
+      outPopCount += BitUtil::kBytePopcount[out[i]];
```

Comment:
This extra line may disable compiler to vectorize the loop and cause big performance regression.[1]
But from your benchmark result, there's at most 20% difference.
Would you list the test names in your benchmark result graph? BitmapAnd related benchmark is about 9G/s on my test machine (no faster than most desktop pc), looks your test result is about 400M/s.

[1] https://quick-bench.com/q/ko7aKpF3R2N8G2yqN3g5mUKgOPg

Reply:
I have checked it with clang shouldn't be different with gcc/msvc. Here is the godbolt snippet with vectorization flags showing same type of vectorization happening with pop-count and without. https://godbolt.org/z/bGhdGEjr6

Yeah, M/s is my mistake. Those are numbers from debug build I did, I forgot to switch to release. With release, the  numbers are much higher. I shall update the description with new numbers soon and with benchmark labels.

Reply:
> Here is the godbolt snippet with vectorization flags showing same type of vectorization happening with pop-count and without. https://godbolt.org/z/bGhdGEjr6 But I will check with gcc/msvc too

In your test code, line 34~43.
The problem is that `v` is calculated but never used. So compiler will simply delete all the code calculating it.
Change the function to let it return `v`, the generated code will be totally different.

```cpp
void WithPopcount(const uint8_t* a, const uint8_t* b, const int N) {
  std::vector<uint8_t> c;
  c.reserve(N);

  unsigned int v = 0;
  for (int i = 0; i < N; ++i) {
    c[i] = a[i] & b[i];
    v += table[c[i]];
  }
}
```

> Yeah, M/s is my mistake. Those are numbers from debug build I did, I forgot to switch to release. With release, the numbers are much higher. I shall update the description with new numbers soon and with benchmark labels.

No worry, it happens.

Reply:
Upon quick check, similar isn't observed on gcc and msvc. Will push updates once they are ready. Thank you for the feedback!

Reply:
> The problem is that v is calculated but never used. So compiler will simply delete all the code calculating it.
> Change the function to let it return v, the generated code will be totally different.

Makes sense. After playing with it for few minutes, saving result to register and then using that for other two calculations is what letting compiler vectorize the loop on all three compilers consistently.

- clang - https://godbolt.org/z/sKvodGzbr
- gcc - https://godbolt.org/z/Kdc38P7fx
- msvc - https://godbolt.org/z/n1so9PbG3

In any case, eventually I dropped the above approach. 

I had a version of CountSetBits earlier that was doing popcount on 64-bit words. But I removed that in favour of lookup during the bitmap op assuming lookup might be faster. It turned out just doing popcount finally on the whole array is more efficient than clubbing the pop-count with bitmapOp .

Please check the updated description for new numbers